File: quicksort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// time complexity: O(nlogn), worst case: O(n^2).

void quicksort(int *arr, int low, int high);
int partition(int *arr, int low, int high);
void swap(int *a, int *b);


int main(void)
{
    srand(time(NULL));
    int arr[] = {4, 5, 6, 3, 2, 1, 7, 8, 9, 10};
    int len = sizeof(arr) / sizeof(arr[0]);
    quicksort(arr, 0, len - 1);
    for (int i = 0; i < len; i++)
    {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

void quicksort(int *arr, int low, int high)
{
    if (low < high)
    {
        int pivotal = partition(arr, low, high);
        quicksort(arr, low, pivotal - 1);
        quicksort(arr, pivotal + 1, high);
    }
}

int partition(int *arr, int low, int high)
{
    int pivotal_idx = rand() % (high - low) + low;
    int pivotal = arr[pivotal_idx];

    swap(&arr[pivotal_idx], &arr[high]);
    
    int i = low;
    for (int j = low; j < high; j++)
    {
        if (arr[j] <= pivotal)
        {
            swap(&arr[i], &arr[j]);
            i++;
        }
    }
    swap(&arr[high], &arr[i]);
    return i;

}

void swap(int *a, int *b)
{
    int tmp = *a;
    *a = *b;
    *b = tmp;
}

File: quicksort.py
from random import randint


def main():
    arr = [2, 4, 5, 36, 3, 5, 45, 12, 64, 23, 1, 0, 9, 9]
    quickSort(arr, 0, len(arr) - 1)
    print(arr)

def swap(list, a, b):
    list[a], list[b] = list[b], list[a]
    return list


def partition(arr, low, high):
    pivotal_idx = randint(low, high)
    swap(arr, pivotal_idx, high)
    i  = low
    for j in range(low, high):
        if arr[j] < arr[high]:
            swap(arr, i, j)
            i += 1
    swap(arr, i, high)
    return i



def quickSort(arr, low, high):
    if low < high:
        pivotal_idx = partition(arr, low, high)
        quickSort(arr, low, pivotal_idx - 1)
        quickSort(arr, pivotal_idx + 1, high)

main()